samza-core/src/main/java/org/apache/samza/clustermanager/SamzaApplicationState.java

```diff
@@ -115,6 +115,13 @@
    */
   public final ConcurrentHashMap<String, SamzaResourceStatus> failedProcessors = new ConcurrentHashMap<>(0);
 
+
+  /**
+   *  Map of the Samza processor ID to the count of failed attempts
+   *  Modified by AMRMCallbackThread
+   */
+  public final ConcurrentMap<String, AtomicInteger> perProcessorFailureCount = new ConcurrentHashMap<>(0);
+
```

Comment:
If this information is only useful for metric-emission, does it need to be stored in "state" ?

Reply:
That is correct, we can directly wire metrics registry in ContainerManager, ContainerAllocator and instantiate new guage and counters in the code but all metrics related to AM are under this ContainerProcessManagerMetrics class which holds MetricsRegistry and SamzaApplicationState, so once does not need to wire MetricsRegistry individually to each AM class ContainerManager, ContainerAllocator. This is the justification for maintained this state variable to wire metrics, I feel this approach is cleaner

Reply:
Check SamzaApplicationState most of the state there is just used for metric emissions